Add VFS recursive ls for LocalFS, MemFS, and S3

shaunrd0 opened the pull request:
Adds VFS recursive ls for Local, MemFS, and S3.

Initially I defined `Win::ls_recursive`, `Posix::ls_recursive`, but these definitions ended up being identical so in the last commit I merged them into `VFS::ls_recursive`. If we'd rather keep them separate I can revert this commit.

---
TYPE: FEATURE
DESC: Adds VFS recursive ls for Local, MemFS, and S3.

Shelnutt2 commented on tiledb/sm/filesystem/vfs.h:
```diff
@@ -348,6 +356,16 @@ class VFS {
+   * @param uris The URIs that are contained in the parent.
+   * @param max_paths The maximum number of paths to be retrieved.
+   */
+  void ls_recursive(
```
Should we have a `optional<std::vector<URI>>` return instead of pass the vector as a pointer?

shaunrd0 replied:
Yep that makes sense. These will throw on failure though, unless I'm mistaken we'd never return the nullopt so I updated to return `vector<URI>`.

davisp commented on tiledb/api/c_api/vfs/vfs_api.cc:
```diff
@@ -297,6 +297,30 @@ capi_return_t tiledb_vfs_ls(
+  auto children = vfs->ls_recursive(tiledb::sm::URI(path), max_paths);
+
+  // Apply first callback with empty path to set first offset of 0.
+  int rc = callback("", data, data_offsets);
```
This should go. You're coordinating your callback implementation with the callback API. If the callback needs to know its first invocation it should track that data itself or the callback API should be changed dramatically to have a type of invocation (though I don't think that is needed here).

davisp commented on test/src/unit-cppapi-vfs.cc:
```diff
@@ -501,3 +502,79 @@ TEST_CASE(
+    }
+
+    // If max_paths is -1 all results should be returned.
+    // S3 won't return prefixes without objects. +3 for posix d1/, d2/, and d3/.
```
Having an API that only sometimes returns directories seems bug prone. I'd either change it to only return files for the memfs/file URIs or do the leg work processing s3 results to insert directory entries. I'd personally lean towards files only at this point, but only narrowly since it seems easier and would have fewer subtleties on what is and isn't a directory in S3 land.

davisp reviewed (changes requested):
Looks pretty good overall. That callback api/invocation behavior needs changing. I'd slightly prefer to add the sizes to the new `ls` API but that's less important since we can always add it later.

eric-hughes-tiledb commented on test/src/unit-capi-vfs.cc:
```diff
@@ -970,3 +979,89 @@ TEST_CASE_METHOD(VFSFx, "C API: Test VFS parallel I/O", "[capi][vfs]") {
+  tiledb_vfs_t* vfs;
+  auto fs_vec = vfs_test_get_fs_vec();
+  REQUIRE(vfs_test_init(fs_vec, &ctx, &vfs).ok());
+  int64_t max_paths = GENERATE(10, 50, -1);
```
This `GENERATE` declaration needs documentation, at minimum.

If `max_paths` argument is < 0, it's invalid. That test should be separated out in argument validation.

shaunrd0 replied:
I had used -1 here to request all results, similar to `VFS::ls_with_sizes`. I missed writing some docstrings for this parameter though so added them in today. I could use 0 instead with uin64_t, if that's preferred.

I got to most of the feedback here today, aside from the unresolved comments related to testing.

eric-hughes-tiledb commented on tiledb/sm/cpp_api/vfs.h:
```diff
@@ -560,6 +589,23 @@ class VFS {
 
+  /**
+   * Callback function to be used when invoking the C TileDB function
+   * for recursive ls.
```
It's clearly a callback. What does the callback do?

shaunrd0 commented on tiledb/api/c_api/vfs/vfs_api_external.h:
```diff
@@ -607,6 +607,48 @@ TILEDB_EXPORT capi_return_t tiledb_vfs_ls(
+    int32_t (*callback)(const char*, void*, void*),
+    void* data,
+    void* data_off,
+    int64_t max_count) TILEDB_NOEXCEPT;
```
I see what you're saying with the second `void*`. Is this still a concern with the change to a single `void*` as a struct? It seems like after this change the caller would handle the overflow in the callback, like I did here:

https://github.com/TileDB-Inc/TileDB/pull/4266/commits/07bd3d7e78c38d57dfc64f8e2deef3cb07bdfadd#diff-63143b90635ecffd1a8c204f366f452e9a75795f4c405846f4195c371448a3adR997

Otherwise internally we're using `std::string` in our `ls_recursive_cb`, so that would reallocate as we append data

shaunrd0 commented on test/src/unit-capi-vfs.cc:
```diff
@@ -970,3 +979,89 @@ TEST_CASE_METHOD(VFSFx, "C API: Test VFS parallel I/O", "[capi][vfs]") {
   }
 }
+
+TEST_CASE("C API: VFS recursive ls", "[capi][vfs][ls-recursive]") {
```
I looked at these tests and I don't see a way to call ls_recursive without VFS using C APIs. It feels not right to add VFS tests to `sm/filesystem`, I could move them to unit_capi_vfs?

davisp reviewed (changes requested):
For the most part, the only thing left I'm seeing is that the comments are out of date compared to what the code does. They're still talking a lot about offsets which doesn't really exist as part of the API any more (client code may elect to track them, and maybe that'd be a useful example though?).

shaunrd0 commented:
Two quick notes, aside from this I addressed all the feedback here

I kept the GENERATE with text prefixes because I don't see a way to GENERATE over VFS objects or tiledb_filesystem_t. S3 and MemFilesystem don't implement ls_recursive and the ladder doesn't define TILEDB_LOCALFS, so using text prefixes is the only way I could come up with to keep this in a single test without duplicating things.

We met and discussed making some changes to introduce ls_v2 and deprecate the current ls. I created a separate story for ls_v2 since this is getting large: https://app.shortcut.com/tiledb-inc/story/33660/deprecate-tiledb-vfs-ls-and-introduce-tiledb-vfs-ls-v2

I'm looking at the standalone mac failure now on my fork.. If anything stands out there LMK, looks like the issue is in `unit_vfs.cc` :+1:

davisp commented on tiledb/sm/cpp_api/vfs_experimental.h:
```diff
@@ -0,0 +1,89 @@
+      const std::string& uri,
+      const tiledb::sm::VFS::LsInclude& predicate =
+          tiledb::sm::VFS::ls_include) {
+    tiledb::sm::VFS::LsRecursiveData ls_data;
```
The definitions in `tiledb::sm::*` are internal and can't be referenced here because this is a public API header.

shaunrd0 replied:
done

eric-hughes-tiledb commented on tiledb/api/c_api/vfs/vfs_api_experimental.h:
```diff
@@ -0,0 +1,101 @@
+ * @param object_size The size of the object at the current path.
+ * @param data Data passed to the callback used to store collected results.
+ */
+typedef int32_t (*LsCallback)(
```
```suggestion
typedef int32_t (*tiledb_ls_callback_t)(
```
This is a definition in an external global namespace, so it needs the standard prefix.

shaunrd0 replied:
done

davisp commented on tiledb/sm/filesystem/posix.cc:
```diff
@@ -299,6 +300,60 @@ tuple<Status, optional<std::vector<directory_entry>>> Posix::ls_with_sizes(
+    // via file_size() and ls() calls will feel this too.
+    // If this penalty becomes noticeable, we should just duplicate
+    // this implementation in ls() and don't get the size
+    if (next_path->d_type == DT_DIR) {
```
```suggestion
    if (recursive && next_path->d_type == DT_DIR) {
```

I'm pretty sure we're missing a check on whether this is a recursive `ls` here.

davisp replied:
Which also means we probably don't have tests for the non-recursive `ls_cb` which we should probably add.

davisp commented on tiledb/sm/filesystem/posix.cc:
```diff
@@ -299,6 +300,60 @@ tuple<Status, optional<std::vector<directory_entry>>> Posix::ls_with_sizes(
+    }
+    std::string abspath = path + "/" + next_path->d_name;
+
+    // Getting the file size here incurs an additional system call
+    // via file_size() and ls() calls will feel this too.
+    // If this penalty becomes noticeable, we should just duplicate
+    // this implementation in ls() and don't get the size
```
I'd probably delete this comment. I know where it comes from and it kind of made sense at the time. Seeing as we haven't found a need to make an optimized version for ls since I'd say the warning is out dated and can be removed.

davisp reviewed (changes requested):
Good work. This all looks great. I just caught one last minor issue on the recursive parameter inside VFS that we'll want to fix and then cover with a test for non-recursive `ls_cb`.

eric-hughes-tiledb commented on tiledb/sm/filesystem/posix.cc:
```diff
@@ -299,6 +300,56 @@ tuple<Status, optional<std::vector<directory_entry>>> Posix::ls_with_sizes(
+  std::string path = uri.to_path();
+  struct dirent** paths;
+  // scandir to traverse sorted objects, so we can invoke callback directly.
+  int path_num = scandir(path.c_str(), &paths, NULL, alphasort);
```
A raw call to a function (in this case a system call) that directly allocates memory needs to be wrapped in exception-safe types at the point of call, not afterwards, as is done here. `scandir` allocates ones for each string, which means that we need an exception-safe string class. Nothing in the standard library does this, so we'll have to write one. The absence of such wrapper shows up as raw calls to `free` here and some other ones in test code.

eric-hughes-tiledb reviewed (commented):
I'm neither approving nor denying this code. It's got good code in it but also a lot of code, mostly in tests, that needs improvement. It's had scope creep, which has led to an unfortunate intermingling of all these concerns.

There is exactly one item that IMO blocks release: The documentation for the C API function needs to (1) have a section that explains what's experimental about it, and (2) specifies clearly what platforms the call supports.

Here's a list of things we need to change before we consider the whole feature done. These all apply to the present code, but not to completion of the present PR.
* Copypasta in tests about support for S3
* Opaque test parameters that should be in test-parameter classes, and documented
* Repetitive tests that ought to be table-driven.
* C.41-noncompliant `VfsFixture` and the whole fixture life cycle generally. I'm not convinced at all the the Catch2 fixture is particularly good for these tests. That discussion is out of scope for this PR.
* Undocumented copypasta between integration and unit tests. When we need to repeat tests, they should be at least documented to be the same as something. Ideally they should be the same test code instantiated for each situation as templates.
* Memory management of `scandir`.
